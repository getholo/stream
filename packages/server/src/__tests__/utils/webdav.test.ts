import {
  createFile,
  createFolder,
  CreateFileParams,
  createFolderParams,
  createXML,
} from '~/utils/webdav';

describe('Webdav formatting', () => {
  describe('Format a file', () => {
    it('When given all parameters, then the webdav should contain all fields', () => {
      // Arrange
      const input: CreateFileParams = {
        mimeType: 'video/x-matroska',
        modifiedTime: Date.now(),
        path: '/to/infinity/and/beyond',
        size: 1200,
      };

      // Act
      const output = createFile(input);

      // Assert
      expect(output).toEqual({
        href: input.path,
        propstat: {
          prop: {
            displayname: 'beyond',
            getcontentlength: input.size,
            getcontenttype: input.mimeType,
            getlastmodified: new Date(input.modifiedTime).toUTCString(),
          },
          status: 'HTTP/1.1 200 OK',
        },
      });
    });
  });

  describe('Format a folder', () => {
    it('When given all parameters, then the webdav should contain all fields', () => {
      // Arrange
      const input: createFolderParams = {
        path: '/to/infinity/and/beyond',
      };

      // Act
      const output = createFolder(input);

      // Assert
      expect(output).toEqual({
        href: input.path,
        propstat: {
          prop: {
            displayname: 'beyond',
            resourcetype: {
              collection: {},
            },
          },
          status: 'HTTP/1.1 200 OK',
        },
      });
    });
  });

  describe('Format to XML', () => {
    it('When given xml object, then actual xml should be valid', () => {
      // Arrange
      const input = [
        createFolder({
          path: '/sup',
        }),
        createFile({
          mimeType: 'video/mp4',
          modifiedTime: 1572985258000,
          path: '/sup/not much',
          size: 1200,
        }),
      ];

      // Act
      const output = createXML(input);

      // Assert
      expect(output).toMatchSnapshot();
    });
  });
});
