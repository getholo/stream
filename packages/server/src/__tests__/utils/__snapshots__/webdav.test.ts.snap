// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webdav formatting Format to XML When given xml object, then actual xml should be valid 1`] = `"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?><d:multistatus xmlns:d=\\"DAV:\\"><d:response><d:href>/sup</d:href><d:propstat><d:prop><d:displayname>sup</d:displayname><d:resourcetype><d:collection/></d:resourcetype></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response><d:response><d:href>/sup/not much</d:href><d:propstat><d:prop><d:displayname>not much</d:displayname><d:getcontentlength>1200</d:getcontentlength><d:getcontenttype>video/mp4</d:getcontenttype><d:getlastmodified>Tue, 05 Nov 2019 20:20:58 GMT</d:getlastmodified></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response></d:multistatus>"`;
